docs/dev/topology-over-raft.md

```diff
@@ -318,6 +324,32 @@ This happens on tablet migration.
 
 Unlike with static sharding, consecutive tokens are not owned by consecutive shards (modulo shard count).
 
+## Shard assignment stability
+
+When tablet is not in transition, each host may contain at most one tablet replica, so there is a single shard for a given
+token and tablet sharder returns that shard, or shard 0 if there is no replica (for consistency with the current API).
+
+The sharder reports a given shard to be the owning shard for a given token as long as either the previous or next
+replica set has replica on that shard.
+
+This is necessary regardless of what the current read or write selectors in the tablet_transition_info
+are. The coordinator may use a different version of effective_replication_map. It may route read request to the leaving
+replica when the leaving replica already sees the write_both_read_new stage. The read should still be served successfully
+from the leaving tablet replica. Because of that, sharder responses should be stable throughout transition as to not
+cause discrepancy between the coordinator-side view of topology and the replica-side view.
+During transitions which are not intra-node migrations the coordinator decisions about target replica set may vary,
+affected by read and write selectors, but replica-side decisions about shard ownership are constant.
+
```

Comment:
I am not sure I understand what you mean by this paragraph. A coordinator may direct a read to an old replica (in this case to an old shard), but how replica knows what shard the read was sent to? If it is already in `write_both_read_new ` it will direct the read to the new shard, but this is fine.

Reply:
That paragraph is about "transitions which are not intra-node migrations", so before this PR. In that case, replica-side routing is not affected by read selector, the shard owning the tablet is constant during transition, and depends only on the host. If it's a leaving host, we will use the leaving shard even if selector is write_both_read_new. This is correct and desirable, because even if the selector is write_both_read_new on the replica, the coordinator may still be at "streaming stage (so read_old selector), before the barrier is finished. And we want to serve such a read.

Reply:
In case of intra-node migrations (introduced in this PR), coordinator doesn't see that there is a transition, it still sends to hosts. Replica is doing the routing to shards based on read/write selectors.